---
title: |-
  R Crash Course \
   Part 4 -- Intro to Regex
author: "Rob Colautti"
---


# 1. Overview

Regular Expressions

* aka 'regex' and 'regexp'

* a sort of find-and-replace for nerds

* one of the most powerful data tools I have ever learned

* requires patience and lots of **practice**


# 2. Basic Regex Functions in R

## Four Basic Regex Functions in R:

`grep()` and `grepl()` are equivalent to 'find' in your favorite word processor

* General form: `grep("find this", in.this.object)`

`sub()` and `gsub()` are equivalent to 'find and replace'

* General form: `grep("find this", "replace with this", in.this.object)`

## Three Advanced Regex Functions in R:

`regexpr()` provides more detailed info about the first match

`gregexpr()` provides more detailed results about all matches

More examples [here](https://bioinfomagician.wordpress.com/2013/11/03/regular-expression-tutorial-2-commands-in-r/)

## Start with a simple data frame of species names:

```{r}
Species<-c("petiolata", "verticillatus", "salicaria", "minor")
print(Species)
```

## `grep()` -- returns cell addresses matching query 

```{r}
grep("a",Species)
```

## `grepl()` -- returns T/F associated with 

```{r}
grepl("a",Species)
```

## `sub()` -- replaces first match (in each cell)

```{r}
sub("l","L",Species)
```

## `gsub()`  -- replaces all matches

```{r}
gsub("l","L",Species)
```


# 3. Wildcards

## `\` escape character

## `\\` in R, double-escape is usually needed (because first `\` is used to escape special characters in R, and `\` is itself a special character that needs to be escaped to pass through the function")

## `\\w`= all letters and digits (aka 'words')
```{r}
gsub("w","*","...which 1-100 words get replaced?")
gsub("\\w","*","...which 1-100 words get replaced?")
```

## `\\W` = non-word and non-number (inverse of `\\w`)
```{r}
gsub("\\W","*","...which 1-100  words get replaced?")
```

## `\\s` = spaces
```{r}
gsub("\\s","*","...which 1-100  words get replaced?")
```

## `\\t` = tab character (useful for tab-delimited data)
```{r}
gsub("\\t","*","...which 1-100  words get replaced?")
```

## `\\d` = digits
```{r}
gsub("\\d","*","...which 1-100  words get replaced?")
```

## `\\D` = non-digits
```{r}
gsub("\\D","*","...which 1-100  words get replaced?")
```

## __Two more special wildcards:__

## `\\r` = carriage return

## `\\n` = newline character

Unix/Mac files -- lines usually end with `\\n` only

Windows/DOS files -- lines usually end with `\\r\\n`


# 4. Special characters:

## `|`= 'or'

Example, look for w or e
```{r}
gsub("w|e","*","...which 1-100  words get replaced?")
```

## `.` = any character except new line
```{r}
gsub(".","*","...which 1-100  words get replaced?")
```

So how to search for a period? Use the escape character
```{r}
gsub("\\.","*","...which 1-100  words get replaced?")
```

## Use *, ? + and {} for more complicated searches

## Look at these examples carefully
```{r}
sub("\\w","*","...which 1-100 words get replaced?")
gsub("\\w","*","...which 1-100 words get replaced?")
```

## `+` = 1 or more matches
```{r}
sub("\\w+","*","...which 1-100 words get replaced?")
gsub("\\w+","*","...which 1-100 words get replaced?")
```


## `?` is a 'lazy' search (find the first occurrance)
```{r}
sub("\\w?","*","...which 1-100 words get replaced?")
gsub("\\w?","*","...which 1-100 words get replaced?")
```

## `*` is a 'greedy' search (find the last occurrence)
```{r}
sub("\\w*","*","...which 1-100 words get replaced?")
gsub("\\w*","*","...which 1-100 words get replaced?")
```

## `{n,m}` = between n to m matches
```{r}
gsub("\\w{3,4}","*","...which 1-100 words get replaced?")
```

## `{n}` = exactly n matches
```{r}
gsub("\\w{3}","*","...which 1-100 words get replaced?")
```

## `{n,}`= n or more matches
```{r}
gsub("\\w{4,}","*","...which 1-100 words get replaced?")
```


# 5. List range of options using []

## Find everything in square brackets
```{r}
gsub("[aceihw-z]","*","...which 1-100 words get replaced?")
```

## Find everything in square brackets occurring 1 or more times
```{r}
gsub("[aceihw-z]+","*","...which 1-100 words get replaced?")
```

# 6. ^Start and end of line$

## ^ Start of line

Find species starting with "s"
```{r}
grep("^s",Species)
```

### IMPORTANT: ^ Also 'negates' when used within []

Find species containing any letter other than s
```{r}
grep("[^a]",Species)
```

Replace every letter except s
```{r}
gsub("[^a]","*",Species)
```

## $ End of line

Find species ending with "a"
```{r}
grep("a$",Species)
```


# 7. Capture text with brackets `()`

Capture text using `()` and reprint using `\\1`, `\\2`, etc

Replace each word with its first letter
```{r}
gsub("(\\w)\\w+","\\1","...which 1-100 words get replaced?")
```

Pull out only the numbers and reverse their order
```{r}
gsub(".*([0-9]+)-([0-9]+).*","\\2-\\1","...which 1-100 words get replaced?")
```

Reverse first two letters of each word
```{r}
gsub("(\\w)(\\w)(\\w+)","\\2\\1\\3","...which 1-100 words get replaced?")
```

# 8. 'Scraping' data from online sources

Use `readLines` with `curl()` to 'scrape' data from the internet, then `grep()` and `gsub()` into useful structure:
```{r}
library(curl)
Prot<-readLines(curl("http://www.rcsb.org/pdb/files/1ema.pdb"))
Prot[grep("TITLE",Prot)]
```


# EXAMPLE 1: Genbank Record
Download the [Lythrum 18S rRNA gene](https://www.ncbi.nlm.nih.gov/nuccore/7595475/) as a text file, and read it into an object called Lythrum.18S
```{r,echo=F}
Lythrum.18S<-"LOCUS       AF206955                1740 bp    DNA     linear   PLN 18-APR-2003
DEFINITION  Lythrum salicaria 18S ribosomal RNA gene, complete sequence.
ACCESSION   AF206955
VERSION     AF206955.1  GI:7595475
KEYWORDS    .
SOURCE      Lythrum salicaria
  ORGANISM  Lythrum salicaria
            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;
            Spermatophyta; Magnoliophyta; eudicotyledons; Gunneridae;
            Pentapetalae; rosids; malvids; Myrtales; Lythraceae; Lythrum.
REFERENCE   1  (bases 1 to 1740)
  AUTHORS   Soltis,P.S., Soltis,D.E. and Chase,M.W.
  TITLE     Direct Submission
  JOURNAL   Submitted (19-NOV-1999) School of Biological Sciences, Washington
            State University, Pullman, WA 99164-4236, USA
FEATURES             Location/Qualifiers
     source          1..1740
                     /organism=\"Lythrum salicaria\"
                     /mol_type=\"genomic DNA\"
                     /db_xref=\"taxon:13129\"
                     /note=\"Lythrum salicaria L.\"
     rRNA            1..1740
                     /product=\"18S ribosomal RNA\"
ORIGIN      
        1 gtcatatgct tgtctcaaag attaagccat gcatgtgtaa gtatgaacaa attcagactg
       61 tgaaactgcg aatggctcat taaatcagtt atagtttgtt tgatggtatc tgctactcgg
      121 ataaccgtag taattctaga gctaatacgt gcaacaaacc ccgacttctg gaagggacgc
      181 atttattaga taaaaggtcg acgcgggctt tgcccgatgc tctgatgatt catgataact
      241 tgacggatcg cacggccatc gtgccggcga cgcatcattc aaatttctgc cctatcaact
      301 ttcgatggta ggatagtggc ctaccatggt gtttacgggt aacggagaat tagggttcga
      361 ttccggagag ggagcctgag aaacggctac cacatccaag gaaggcagca ggcgcgcaaa
      421 ttacccaatc ctgacacggg gaggtagtga caataaataa caatactggg ctctttgagt
      481 ctggtaattg gaatgagtac aatctaaatc ccttaacgag gatccattgg agggcaagtc
      541 tggtgccagc agccgcggta attccagctc caatagcgta tatttaagtt gttgcagtta
      601 aaaagctcgt agttggacct tgggttgggt cgaccggtcc gcctttggtg tgcaccgatc
      661 ggctcgtccc ttctaccggc gatgcgcgcc tggccttaat tggccgggtc gttcctccgg
      721 tgctgttact ttgaagaaat tagagtgctc aaagcaagca ttagctatga atacattagc
      781 atgggataac attataggat tccgatccta ttatgttggc cttcgggatc ggagtaatga
      841 ttaacaggga cagtcggggg cattcgtatt tcatagtcag aggtgaaatt cttggattta
      901 tgaaagacga acaactgcga aagcatttgc caaggatgtt ttcattaatc aagaacgaaa
      961 gttgggggct cgaagacgat cagataccgt cctagtctca accataaacg atgccgacca
     1021 gggatcagcg aatgttactt ttaggacttc gctggcacct tatgagaaat caaagttttt
     1081 gggttccggg gggagtatgg tcgcaaggct gaaacttaaa ggaattgacg gaagggcacc
     1141 accaggagtg gagcctgcgg cttaatttga ctcaacacgg ggaaacttac caggtccaga
     1201 catagtaagg attgacagac tgagagctct ttcttgattc tatgggtggt ggtgcatggc
     1261 cgttcttagt tggtggagcg atttgtctgg ttaattccgt taacgaacga gacctcagcc
     1321 tgctaactag ctatgtggag gtacacctcc acggccagct tcttagaggg actatggccg
     1381 cttaggccaa ggaagtttga ggcaataaca ggtctgtgat gcccttagat gttctgggcc
     1441 gcacgcgcgc tacactgatg tattcaacga gtctatagcc ttggccgaca ggcccgggta
     1501 atctttgaaa tttcatcgtg atggggatag atcattgcaa ttgttggtct tcaacgagga
     1561 attcctagta agcgcgagtc atcagctcgc gttgactacg tccctgccct ttgtacacac
     1621 cgcccgtcgc tcctaccgat tgaatggtcc ggtgaaatgt tcggatcgcg gcgacgtggg
     1681 cgcttcgtcg ccgacgacgt cgcgagaagt ccattgaacc ttatcattta gaggaaggag
//"
```

The 18S Sequence for Lythrum salicaria (from Genbank)
```{r,echo=FALSE}
cat(Lythrum.18S)
```

Remember our `\\n` newline? Let's use it to delete everything after the first line
```{r}
gsub("\\n.*","",Lythrum.18S)
```

Or use the field names to isolate them:
```{r}
gsub(".*ORIGIN","",Lythrum.18S)
```

The output is a bit messy because R is printing the newline character \\n instead of adding a new line. 

You can use the cat() function to make it look better:
```{r}
cat(gsub(".*ORIGIN","",Lythrum.18S))
```

Let's clean up the sequence. An easy way is to make an object and do it in steps
```{r}
# Isolate the lines containing sequence
Seq<-gsub(".*ORIGIN","",Lythrum.18S)
print(Seq)
# Remove line numbers, newline characters, spaces and /
Seq<-gsub("[\n0-9 /]+","",Seq) ## Note the single \ in this case since it is inside the square brackets
print(Seq)
```

Capital letters look nicer, but requires a PERL qualifier `\\U` that is not standard in R
```{r}
Seq<-gsub("([actg])","\\U\\1",Seq,perl=T)
print(Seq)
```

Look for start codons?
```{r}
gsub("ATG","-->START<--",Seq)
```

Open reading frames?
```{r}
gsub("(ATG([ATGC]{3})+?(TAA|TAG|TGA)+?)","-->\\1<--",Seq)
```

Or go back, and keep non-reading frame in lower case
```{r}
Seq<-gsub(".*ORIGIN","",Lythrum.18S)
Seq<-gsub("[\n0-9 /]+","",Seq) ## Note the single \ in this case since it is inside the square brackets
gsub("(atg([atgc]{3})+?(taa|tag|tga)+?)","\\U\\1",Seq,perl=T)
```


# EXAMPLE 2: Organizing Data
Imagine you have a repeated measures design. 3 transects (A-C) and 3 positions along each transect (1-3)
```{r}
Transect<-data.frame(Species=1:20,A1=rnorm(20),A2=rnorm(20),A3=rnorm(20),B1=rnorm(20),B2=rnorm(20),B3=rnorm(20),C1=rnorm(20),C2=rnorm(20),C3=rnorm(20))
head(Transect)
```

You want to look at only transect A for the first 3 species
```{r}
Transect[1:3,grep("A",names(Transect))]
```

Or look at the first position of each transect for the first 3 species
```{r}
Transect[1:3,grep("1",names(Transect))]
```

Or rows with species IDs containing the number 2
```{r}
Transect[grep("2",Transect$Species),]
```

# PRACTICE EXERCISES

## 1. Consider a vector of email addresses scraped from the internet:

  * robert 'dot' colautti 'at' queensu 'dot' ca
  * chris.eckert[at]queensu.ca
  * lonnie.aarssen at queensu.ca

Use regular expressions to convert all email addresses to the standard format: name@queensu.ca

## 2. Create a random sequence of DNA:
```{r,results="hide"}
My.Seq<-sample(c("A","T","G","C"),1000,replace=T)
```

    * Replace T with U
    * Find all start codons (AUG) and stop codons (UAA, UAG, UGA)
    * Find all open reading frames (hint: consider each sequence beginning with AUG and ending with a stop codon; how do you know if both sequences are in the same reading frame?)
    * Count the length of bp for all open reading frames

## 3. More online examples

[http://regex.sketchengine.co.uk/extra_regexps.html](http://regex.sketchengine.co.uk/extra_regexps.html)

## 4. Regex Golf

Have fun! [LINK](https://alf.nu/RegexGolf)

